### ISSUE TITLE
einsum for SparseTensors

### ISSUE THREAD

**Issue by carlosuriarte** [tensorflow/tensorflow#43497]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac
- TensorFlow installed from (source or binary): Source
- TensorFlow version (or github SHA if from source): 2.2.0

**Provide the text output from tflite_convert**
I do not find an 'optimized' way to compute a multiplication between an sparse and a dense tensor that are not matrices.
For example, I would like to perform the following multiplication between a 3D sparse tensor A and a 2D dense tensor B:
C[i,j] = \sum_k B[i,k] A[i,k,j]

IMPORTANT! A needs to maintain its sparsity. Functions like tf.sparse.to_dense() should be not must not be used!
```
# Example
# inputs
A = tf.sparse.SparseTensor(values=tf.constant([1,1,1,1,2,1,1,1,1,1,-1], tf.float32),indices=[[0,0,1],[0,1,2],[0,2,0],[1,0,0],[1,1,1],[1,1,2],[1,2,2],[2,0,2],[2,1,1],[2,2,1],[2,2,2]], dense_shape=[3,3,3])
B = tf.constant([[1, 2, 3],
                 [4, 5, 6],
                 [7, 8, 9]], tf.float32)

# output
C = tf.constant([[3, 1, 2],
                 [4, 10, 11],
                 [9, 8, -1]], tf.float32)
```
A 'tf.sparse.einsum(B,A, 'ik,ikj->ij')'-like function would be nice to reproduce the behaviour of einsum for dense matrices.

**Comment by abattery** [comment]
FYI, the tf-nightly version supports the following einsum parameters natively during tflite conversion:
IJK,IKM->IJM
BFH,HO->BFO
LBH,BL->BH
LBH,BKL->BKH
BFD,DNH->BFNH
BFND,NDH->BFH
BFNH,BTNH->BNFT
BIN,BINJ->BIJ
BIJ,BINJ->BIN
acebe,aecd->abcd

If your case is not on the list, you can still support your einsum op via Select TF ops.

**Comment by carlosuriarte** [comment]
Hello,

The problem I have is with respect to the sparsity. I need to maintain the sparsity of the tensor in the example.

Could you, please, give me a piece of code of the einsum-like function working for a multiplication with a sparse tensor involved? For example, with A: a 3D tf.sparse.SparseTensor (sparse); B: a 2D tf.constant (dense); I want to obtain C: a 2D tf.constant (dense) such that C[i,j] = \sum_k B[i,k] * A[i,k,j].

If everything was dense it would work as: C = tf.einsum(B,A,'ik,ikj -> ij’)… But it is not the case ;(. I cannot convert A to a dense tensor!

Thank you in advance,

Regards,

Carlos Uriarte Baranda
Email. carlos.uribar@gmail.com
Tel. +34 688 610 487

**Comment by abattery** [comment]
Sorry for encountering this issue. I tried to help you as a TFLite team member. Are you willing to execute this case in TF or TFLite?

**Comment by carlosuriarte** [comment]
In TF. I prefer to avoid TFLite. Maybe I am in the wrong issue branch. Do you know where should I forward this issue?

Regards,

**Comment by abattery** [comment]
@curiarteb This is the right place for the tensorflow issue like this one. The "comp:lite" tag took my attention, which is used for TFLite issues.

**Comment by curiarteb** [comment]
Ok. Sorry for the tag. Any ideas on how can I employ einsum with sparse Tensors? Is it possible in the current TF version?

Carlos Uriarte Baranda
Email. carlos.uribar@gmail.com
Tel. +34 688 610 487

**Comment by curiarteb** [comment]
Any news?

**Comment by sachinprasadhs** [comment]
Have you tried with [tf.sparse.sparse_dense_matmul](https://www.tensorflow.org/api_docs/python/tf/sparse/sparse_dense_matmul) which will allow multiplication between sparse and dense matrix.

**Comment by curiarteb** [comment]
Yes! Of course! The problem is that I need to compute a "multiplication" between a 3D sparse tensor (batch of sparse matrices) and a 2D dense tensor (batch of vectors). The tf.sparse.sparse_dense_matmul function only permits multiplications between 2D sparse/dense tensors. Currently, I reshape the 3D sparse tensor to a 2D huge non-overlapping block matrix, and the 2D tensor to a large 1D tensor (that thereafter I reshape it again to a 2D tensor). The problem on doing this is that I need to access the shape of the 3D sparse tensor. For the fitting function in Keras this is conflicting because a priori the first axis is assumed to be 'None'. This leads us to miss the Keras batch flow behaviour during the training of the network that deals with the aforementioned kind of tensors, forcing us to design a customized and significantly slower low-level training.

I guess a kind of a tf.einsum function dedicated for the sparse library, which currently does not exist, would resolve this issue.

**Comment by MPKenning** [comment]
Is there interest in a sparse einsum? I have an interest in it myself. I want to multiply two matrices (k, v, v) and (k, v, f) together, the first of which is highly sparse, and so far I have found no way to do this efficiently. I have tried using a for-loop over the dimension k and stack the results, but it is neither pretty nor efficient.

**Comment by curiarteb** [comment]
Dear Michael,

I had an equivalent issue multiplying a sparse tensor (k,i,j) by a dense matrix (u,v). I "solved" it doing a for-loop in the k index, but it was extremely inefficient. When iterating along the first axis, I was forced to miss the graph execution mode (the fit function of Keras), and to optimize in eager mode. It provided extremely slow executions ;(.

I am awaiting for Tensorflow to fix it!

I share you my already published work where I found the aforementioned difficulty: https://www.sciencedirect.com/science/article/pii/S0045782521007374

Thank you. Best regards,

**Comment by MPKenning** [comment]
The other difficulty is that stacking the results of a for-loop is apparently forbidden by graph-execution mode: the interpreter complains of a matrix with changing dimensions, which is addressed by relaxing the shapes, but it is not a pretty solution nevertheless and certainly not efficient, as you say. 

This would be such a helpful innovation if it were implemented.

**Comment by Arka161** [comment]
Commenting on this as this looks like a needed feature which would help working with sparse tensors without having to convert them back to a dense format.

**Comment by mikeymezher** [comment]
Though this would definitely be a nice feature to have, I'm pretty confident there is a work around that avoids for loops. 
Consider sparse tensor A of size (b x s x d) and dense tensor B of size (b x k x d)
If we'd like to matrix multiply A x B.T, and expect the result E to be of size (b x s x k), we can do this using sparse_dense_matmul by:  

1. Reshaping A into (b*s x d) and B into (b*k x d)
2. Conduct sparse_dense_matmul, resulting in a tensor R of shape (b*s x b*k)
3. Use gather_nd to extract the necessary elements from R. 

R should contain all elements in E (plus extraneous dot products), and it should be relatively straight forward to determine their corresponding indices. 

This approach does have the potential to be somewhat memory inefficient however, especially if the batch size is large.

**Comment by curiarteb** [comment]
Yes! And it was exactly what I did. However, there are two big issues regarding this:
1) You need to know the size of "b" for reshaping both A and B.
2) The .fit function of Keras enables graph execution mode (it is much faster than eager execution). When constructing the graph,the reshape function breaks, since we are feeding tensors with "None" first axis dimensions.

Then, your reshaping only works in eager mode, which is much slower than graph mode.

**Comment by ruler501** [comment]
Use tf.shape(tensor) to get the shape at runtime in graph mode. As long as the rank is determined working with it is the same.

**Comment by tomoleary** [comment]
> Is there interest in a sparse einsum? I have an interest in it myself. I want to multiply two matrices (k, v, v) and (k, v, f) together, the first of which is highly sparse, and so far I have found no way to do this efficiently. I have tried using a for-loop over the dimension k and stack the results, but it is neither pretty nor efficient.

I am also interested in a sparse einsum!

**Comment by kerighan** [comment]
Any update on this? This feature would be amazing.

**Comment by curiarteb** [comment]
> Any update on this? This feature would be amazing.

It definitely would be! Unfortunately, there doesn't seem to be any update yet ;(.

**Comment by curiarteb** [comment]
Tensorflow still has not developed sufficiently effective and robust tools for the handling and operation of sparse tensors, as in the case described in this issue.

**Comment by Kayyuri** [comment]
Hi @curiarteb,

Thank you so much for filing this issue! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
